### ISSUE TITLE
Remove Unnecessary Cast: Removing boxing conversion in conditional expression breaks code

### ISSUE THREAD

**Issue by DustinCampbell** [dotnet/roslyn#5314]
```C#
using System;

class Program
{
    static void Main(string[] args)
    {
        var x = Convert(true, 42);
        Console.WriteLine(x.GetType());
    }

    static object Convert(bool cond, ulong value)
    {
        return cond ? (object)(uint)value : value;
    }
}
```

The above program should output `System.UInt32`. If the cast to `(object)` is removed, it prints `System.UInt64`.

**Comment by AdamSpeight2008** [comment]
@DustinCampbell Isn't the `var` making the type the smallest type that can fit both options, this case *uint* and *ulong* so picks *ulong* to be the dominant type.

**Comment by AdamSpeight2008** [comment]
The "bug" is within the inline-if (`?:`) as the following expansion works find.
```c#
    static object Convert(bool cond, ulong value)

    {
      if(cond) return (uint)value;
      return value;
//        return cond ? ((uint)value) : value;
    }
```

**Comment by DustinCampbell** [comment]
@AdamSpeight2008: Thanks for posting. I *do* understand the bug and the correct fix, which is in the remove unnecessary cast code that handles conditional expressions.

**Comment by AdamSpeight2008** [comment]
Sorry @DustinCampbell it reads like that, just wasn't clear that the "bug" is in the **Redundant Cast Analyser**

**Comment by DustinCampbell** [comment]
Fixed with PR #5319
